src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -294,56 +398,70 @@ private TimeSpan GetTimeout()
-
-                        _status = new Status(StatusCode.DeadlineExceeded, "Deadline Exceeded");
+                return;
+            }
 
-                        try
-                        {
-                            // TODO(JamesNK): I believe this sends a RST_STREAM with INTERNAL_ERROR. Grpc.Core sends NO_ERROR
-                            HttpContext.Abort();
-                        }
-                        catch (Exception ex)
-                        {
-                            Log.DeadlineCancellationError(_logger, ex);
-                        }
-                    }
-                }
+            // Request abort uses the same cancellation token
+            // Don't run deadline logic if the request has already been aborted
+            if (HttpContext.RequestAborted.IsCancellationRequested)
+            {
+                return;
             }
-        }
 
-        public void Dispose()
-        {
-            var lockTaken = false;
+            Debug.Assert(_deadlineLock != null, "Lock has not been created.");
+
+            await _deadlineLock.WaitAsync();
+
             try
             {
-                // Only take a lock if there is a deadline
-                if (_deadlineTimer != null)
+                // Re-check after lock
+                if (_disposed)
+                {
+                    return;
+                }
+
+                Log.DeadlineExceeded(_logger, GetTimeout());
+
+                var status = new Status(StatusCode.DeadlineExceeded, "Deadline Exceeded");
+
+                var trailersDestination = GrpcProtocolHelpers.GetTrailersDestination(HttpContext.Response);
+                GrpcProtocolHelpers.SetStatus(trailersDestination, status);
+
+                _status = status;
+
+                // Immediately send remaining response content and trailers
+                // If feature is null then abort will still end request, but response won't have trailers
+                var completionFeature = HttpContext.Features.Get<IHttpResponseCompletionFeature>();
+                if (completionFeature != null)
                 {
-                    Monitor.Enter(_deadlineTimer, ref lockTaken);
+                    await completionFeature.CompleteAsync();
                 }
 
-                _disposed = true;
-                _deadlineTimer?.Dispose();
+                // TODO(JamesNK): I believe this sends a RST_STREAM with INTERNAL_ERROR. Grpc.Core sends NO_ERROR
+                HttpContext.Abort();
+            }
+            catch (Exception ex)
+            {
+                Log.DeadlineCancellationError(_logger, ex);
             }
             finally
             {
-                if (lockTaken)
-                {
-                    Monitor.Exit(_deadlineTimer);
-                }
+                _deadlineLock.Release();
             }
         }
 
+        public void Dispose()
+        {
+            _disposed = true;
+            _deadlineCts?.Dispose();
```

Comment:
This is always called with _deadlineLock, right?

Reply:
Yes

Reply:
Hmmm, _deadlineLock isn't being disposed. Do you need to dispose SemaphoreSlim?

Reply:
>  Note
> 
> Always call Dispose before you release your last reference to the SemaphoreSlim. Otherwise, the resources it is using will not be freed until the garbage collector calls the SemaphoreSlim object's Finalize method.

https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim.dispose?view=netcore-3.0